rviz_ogre_vendor/CMakeLists.txt

```diff
@@ -120,7 +120,12 @@ macro(build_ogre)
     set(OGRE_CXX_FLAGS "${OGRE_CXX_FLAGS} /w /EHsc")
   elseif(APPLE)
     set(OGRE_CXX_FLAGS "${OGRE_CXX_FLAGS} -std=c++14 -stdlib=libc++ -w")
-    list(APPEND extra_cmake_args "-DCMAKE_OSX_ARCHITECTURES='x86_64'")
+    message("-- Building for CMAKE_OSX_ARCHITECTURES=${CMAKE_OSX_ARCHITECTURES}")
+    if (NOT CMAKE_OSX_ARCHITECTURES)
+      message(FATAL_ERROR "Must specify CMAKE_OSX_ARCHITECTURES for APPLE builds")
```

Comment:
@kmamykin indeed it seems CMake cannot pull off a sensible default :/

Reply:
If I'm understanding this properly though, this would be a major regression in build functionality.  That is, this package would no longer build on macOS at all without specifying `-DCMAKE_OSX_ARCHITECTURES` on the colcon command-line.  Am I understanding that properly?

Reply:
Thanks for the review and comments.

Regarding having to specify cmake parameter `CMAKE_OSX_ARCHITECTURES`:
I agree, ideally it wouldn't be necessary and the build system would just pick a sensible default. But it looks like that ogre build (and may be other vendored packages) somehow manage to default to x86_64 if `CMAKE_OSX_ARCHITECTURES` is not explicitly passed in. I though fixing those packages would be outside of this PR, and can be addressed separately. For now this is a temp solution to make rviz at least build for mac arm64. 

Here is what I tried:

1) Removing any mention of `CMAKE_OSX_ARCHITECTURES` from `rviz_ogre_vendor/CMakeLists.txt` and not passing cmake argument either. Ideally everything should default to the local architecture. But the vendored OGRE somehow builds for x86_64. Here is the cmake config command

```
cd /Users/kmamykin/Projects/ros2-build-on-macOS/ros2_galactic/build/rviz_ogre_vendor/ogre-v1.12.12-prefix/src/ogre-v1.12.12-build && /opt/homebrew/Cellar/cmake/3.20.3/bin/cmake -DOGRE_STATIC:BOOL=OFF -DOGRE_DEPENDENCIES_DIR=/Users/kmamykin/Projects/ros2-build-on-macOS/ros2_galactic/build/rviz_ogre_vendor/ogredeps -DOGRE_INSTALL_PDB:BOOL=OFF -DOGRE_BUILD_DEPENDENCIES:BOOL=OFF -DOGRE_BUILD_TESTS:BOOL=OFF -DOGRE_BUILD_SAMPLES:BOOL=FALSE -DOGRE_INSTALL_SAMPLES:BOOL=FALSE -DOGRE_INSTALL_SAMPLES_SOURCE:BOOL=FALSE -DOGRE_CONFIG_THREADS:STRING=0 -DOGRE_RESOURCEMANAGER_STRICT:STRING=2 -DCMAKE_SKIP_INSTALL_RPATH:BOOL=ON -DCMAKE_INSTALL_PREFIX=/Users/kmamykin/Projects/ros2-build-on-macOS/ros2_galactic/build/rviz_ogre_vendor/ogre_install -DOGRE_BUILD_LIBS_AS_FRAMEWORKS:BOOL=OFF -DOGRE_BUILD_COMPONENT_PYTHON:BOOL=FALSE -DOGRE_BUILD_COMPONENT_JAVA:BOOL=FALSE -DOGRE_BUILD_COMPONENT_CSHARP:BOOL=FALSE -DOGRE_BUILD_COMPONENT_BITES:BOOL=FALSE -DOGRE_ENABLE_PRECOMPILED_HEADERS:BOOL=OFF -DCMAKE_BUILD_TYPE= -DOGRE_BUILD_RENDERSYSTEM_GL:BOOL=TRUE -DOGRE_BUILD_RENDERSYSTEM_D3D11:BOOL=OFF -DOGRE_BUILD_RENDERSYSTEM_D3D9:BOOL=OFF -DCMAKE_C_FLAGS= "-DCMAKE_CXX_FLAGS= -std=c++14 -stdlib=libc++ -w" -DOGRE_CONFIG_ENABLE_ZIP:BOOL=ON -DCMAKE_C_COMPILER=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -DCMAKE_CXX_COMPILER=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -Wno-dev "-GUnix Makefiles" /Users/kmamykin/Projects/ros2-build-on-macOS/ros2_galactic/build/rviz_ogre_vendor/ogre-v1.12.12-prefix/src/ogre-v1.12.12
```
and later all compiles have `-arch x86_64`, which obviously fails to link with arch64 binaries later on.

2) Building universal binaries, passing `CMAKE_OSX_ARCHITECTURES=arm64;x86_64` to cmake args.
`fastcdr` fails to build with 

```
cd /Users/kmamykin/Projects/ros2-build-on-macOS/ros2_galactic/build/fastcdr/src/cpp && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DFASTCDR_SOURCE -Dfastcdr_EXPORTS -I/Users/kmamykin/Projects/ros2-build-on-macOS/ros2_galactic/src/eProsima/Fast-CDR/include -I/Users/kmamykin/Projects/ros2-build-on-macOS/ros2_galactic/build/fastcdr/include -O3 -DNDEBUG -arch arm64 -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk -fPIC -std=c++14 -Wall -Wextra -Wshadow -Wnon-virtual-dtor -pedantic -Wcast-align -Wunused -Woverloaded-virtual -Wconversion -Wsign-conversion -Wdouble-promotion -Wold-style-cast -MD -MT src/cpp/CMakeFiles/fastcdr.dir/exceptions/NotEnoughMemoryException.cpp.o -MF CMakeFiles/fastcdr.dir/exceptions/NotEnoughMemoryException.cpp.o.d -o CMakeFiles/fastcdr.dir/exceptions/NotEnoughMemoryException.cpp.o -c /Users/kmamykin/Projects/ros2-build-on-macOS/ros2_galactic/src/eProsima/Fast-CDR/src/cpp/exceptions/NotEnoughMemoryException.cpp
/Users/kmamykin/Projects/ros2-build-on-macOS/ros2_galactic/src/eProsima/Fast-CDR/src/cpp/FastCdr.cpp:349:2: error: unsupported long double type and no __float128 available
#error unsupported long double type and no __float128 available
 ^
```

Reply:
How about defaulting to x86_64 if `CMAKE_OSX_ARCHITECTURES` is not specified? You'd still be able to tweak it to build on arm64 without breaking everyone else's x86_64 builds. Once the vendored OGRE gets fixed we can drop this.

Reply:
How about trying to use [CMAKE_SYSTEM_PROCESSOR](https://cmake.org/cmake/help/v3.16/variable/CMAKE_SYSTEM_PROCESSOR.html)?  That would at least give us a fighting chance of building it on the M1 automagically.

Reply:
Awesome suggestions. Let me experiment and update the PR with a decent default.